{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "0.data_preparation.ipynb",
      "provenance": [],
      "mount_file_id": "1SXG8S2o47oF8EZXYpjJlIjDLtHl27KwR",
      "authorship_tag": "ABX9TyOpm8Fbijmxv8QxJ4fjiusH",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/yutaro-tanaka-yt2705/ag-cgan/blob/main/0_data_preparation.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "40rjlyCZzTFf"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "from tqdm import tqdm\n",
        "import matplotlib.pyplot as plt\n",
        "import gzip\n",
        "import warnings\n",
        "warnings.filterwarnings(\"ignore\")"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def get_vcf_header(filename):\n",
        "    with gzip.open(filename, \"rb\") as fi:\n",
        "        for l in fi:\n",
        "            l = l.decode(\"utf-8\")\n",
        "            if l.startswith(\"##\"):\n",
        "                continue\n",
        "            elif l.startswith(\"#\"):\n",
        "                return l[1:].strip().split(\"\\t\")\n",
        "            else:\n",
        "                raise ValueError(\"Something wrong in the vcf file!\")\n",
        "\n",
        "#Obtain data from 1000 Genomes. For this, we will be using Chromosome X. \n",
        "vcf_filename = '/content/drive/MyDrive/artificial_genome_project/ALL.chrX.shapeit2_integrated_snvindels_v2a_27022019.GRCh38.phased.vcf.gz'\n",
        "\n",
        "vcf = pd.read_csv(vcf_filename, comment='#', sep=\"\\t\")\n",
        "vcf.columns = get_vcf_header(vcf_filename)\n",
        "print(vcf.shape)\n",
        "vcf.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 317
        },
        "id": "cJdMgiwizVFl",
        "outputId": "f98dd1c0-872a-4ecf-e4a4-fabcaf61755e"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "(106962, 2557)\n"
          ]
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  CHROM    POS ID REF ALT QUAL FILTER  \\\n",
              "0     X  12583  .   C   T    .   PASS   \n",
              "1     X  13587  .   T   C    .   PASS   \n",
              "2     X  13590  .   T   C    .   PASS   \n",
              "3     X  13615  .   A   G    .   PASS   \n",
              "4     X  13878  .   G   C    .   PASS   \n",
              "\n",
              "                                                INFO FORMAT HG00096  ...  \\\n",
              "0  AC=97;AN=5096;DP=8520;AF=0.02;EAS_AF=0;EUR_AF=...     GT     0|0  ...   \n",
              "1  AC=198;AN=5096;DP=5147;AF=0.04;EAS_AF=0.03;EUR...     GT     0|0  ...   \n",
              "2  AC=198;AN=5096;DP=5128;AF=0.04;EAS_AF=0.03;EUR...     GT     0|0  ...   \n",
              "3  AC=20;AN=5096;DP=4673;AF=0;EAS_AF=0;EUR_AF=0;A...     GT     0|0  ...   \n",
              "4  AC=66;AN=5096;DP=3473;AF=0.01;EAS_AF=0;EUR_AF=...     GT     0|0  ...   \n",
              "\n",
              "  NA21128 NA21129 NA21130 NA21133 NA21135 NA21137 NA21141 NA21142 NA21143  \\\n",
              "0     0|0     0|0     0|0     0|0     0|0     0|0     0|0     0|0     0|0   \n",
              "1     0|0     1|0     0|0     0|1     0|0     0|0     0|0     0|0     0|0   \n",
              "2     0|0     1|0     0|0     0|1     0|0     0|0     0|0     0|0     0|0   \n",
              "3     0|0     0|0     0|0     0|0     0|0     0|0     0|0     0|0     0|0   \n",
              "4     0|0     0|0     0|0     0|0     0|0     0|0     0|0     0|0     0|0   \n",
              "\n",
              "  NA21144  \n",
              "0     0|0  \n",
              "1     0|0  \n",
              "2     0|0  \n",
              "3     0|0  \n",
              "4     0|0  \n",
              "\n",
              "[5 rows x 2557 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-5e14527c-5a2e-4c40-a49c-268855047e9a\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>CHROM</th>\n",
              "      <th>POS</th>\n",
              "      <th>ID</th>\n",
              "      <th>REF</th>\n",
              "      <th>ALT</th>\n",
              "      <th>QUAL</th>\n",
              "      <th>FILTER</th>\n",
              "      <th>INFO</th>\n",
              "      <th>FORMAT</th>\n",
              "      <th>HG00096</th>\n",
              "      <th>...</th>\n",
              "      <th>NA21128</th>\n",
              "      <th>NA21129</th>\n",
              "      <th>NA21130</th>\n",
              "      <th>NA21133</th>\n",
              "      <th>NA21135</th>\n",
              "      <th>NA21137</th>\n",
              "      <th>NA21141</th>\n",
              "      <th>NA21142</th>\n",
              "      <th>NA21143</th>\n",
              "      <th>NA21144</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>X</td>\n",
              "      <td>12583</td>\n",
              "      <td>.</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>.</td>\n",
              "      <td>PASS</td>\n",
              "      <td>AC=97;AN=5096;DP=8520;AF=0.02;EAS_AF=0;EUR_AF=...</td>\n",
              "      <td>GT</td>\n",
              "      <td>0|0</td>\n",
              "      <td>...</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>X</td>\n",
              "      <td>13587</td>\n",
              "      <td>.</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>.</td>\n",
              "      <td>PASS</td>\n",
              "      <td>AC=198;AN=5096;DP=5147;AF=0.04;EAS_AF=0.03;EUR...</td>\n",
              "      <td>GT</td>\n",
              "      <td>0|0</td>\n",
              "      <td>...</td>\n",
              "      <td>0|0</td>\n",
              "      <td>1|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|1</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>X</td>\n",
              "      <td>13590</td>\n",
              "      <td>.</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>.</td>\n",
              "      <td>PASS</td>\n",
              "      <td>AC=198;AN=5096;DP=5128;AF=0.04;EAS_AF=0.03;EUR...</td>\n",
              "      <td>GT</td>\n",
              "      <td>0|0</td>\n",
              "      <td>...</td>\n",
              "      <td>0|0</td>\n",
              "      <td>1|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|1</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>X</td>\n",
              "      <td>13615</td>\n",
              "      <td>.</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>.</td>\n",
              "      <td>PASS</td>\n",
              "      <td>AC=20;AN=5096;DP=4673;AF=0;EAS_AF=0;EUR_AF=0;A...</td>\n",
              "      <td>GT</td>\n",
              "      <td>0|0</td>\n",
              "      <td>...</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>X</td>\n",
              "      <td>13878</td>\n",
              "      <td>.</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>.</td>\n",
              "      <td>PASS</td>\n",
              "      <td>AC=66;AN=5096;DP=3473;AF=0.01;EAS_AF=0;EUR_AF=...</td>\n",
              "      <td>GT</td>\n",
              "      <td>0|0</td>\n",
              "      <td>...</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "      <td>0|0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows × 2557 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-5e14527c-5a2e-4c40-a49c-268855047e9a')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-5e14527c-5a2e-4c40-a49c-268855047e9a button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-5e14527c-5a2e-4c40-a49c-268855047e9a');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 2
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#count number of mutations in HG00096 (0:reference allele, 1:alternative allele)\n",
        "vcf.HG00096.value_counts()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DO7V6GnD0wyQ",
        "outputId": "8a6e305f-f3cc-4cda-f62e-2765f088e5f9"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "0|0    98930\n",
              "1|1     2915\n",
              "0|1     2606\n",
              "1|0     2511\n",
              "Name: HG00096, dtype: int64"
            ]
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Prepare table of variant info.\n",
        "variant_info = vcf[['CHROM', 'POS', 'REF', 'ALT', 'INFO']]\n",
        "variant_info['allele_count'] = variant_info.INFO.apply(lambda x: int(x.split(';')[0].replace('AC=', '')))\n",
        "variant_info['allele_freq'] = variant_info.INFO.apply(lambda x: float(x.split(';')[3].replace('AF=', '')))\n",
        "variant_info['eas_allele_freq'] = variant_info.INFO.apply(lambda x: float(x.split(';')[4].replace('EAS_AF=', '')))\n",
        "variant_info['eur_allele_freq'] = variant_info.INFO.apply(lambda x: float(x.split(';')[5].replace('EUR_AF=', '')))\n",
        "variant_info['afr_allele_freq'] = variant_info.INFO.apply(lambda x: float(x.split(';')[6].replace('AFR_AF=', '')))\n",
        "variant_info['amr_allele_freq'] = variant_info.INFO.apply(lambda x: float(x.split(';')[7].replace('AMR_AF=', '')))\n",
        "variant_info['sas_allele_freq'] = variant_info.INFO.apply(lambda x: float(x.split(';')[8].replace('SAS_AF=', '')))\n",
        "variant_info = variant_info.drop('INFO', axis=1)\n",
        "#save table to GDrive.\n",
        "variant_info.to_csv('/content/drive/MyDrive/artificial_genome_project/chrX_variant_info_formatted.csv', index=False)\n",
        "variant_info.head()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "dkMXPs9SztLm",
        "outputId": "f55c8d61-6580-4769-dd84-1600f98fb8b9"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  CHROM    POS REF ALT  allele_count  allele_freq  eas_allele_freq  \\\n",
              "0     X  12583   C   T            97         0.02             0.00   \n",
              "1     X  13587   T   C           198         0.04             0.03   \n",
              "2     X  13590   T   C           198         0.04             0.03   \n",
              "3     X  13615   A   G            20         0.00             0.00   \n",
              "4     X  13878   G   C            66         0.01             0.00   \n",
              "\n",
              "   eur_allele_freq  afr_allele_freq  amr_allele_freq  sas_allele_freq  \n",
              "0             0.00             0.07             0.01             0.00  \n",
              "1             0.02             0.06             0.02             0.05  \n",
              "2             0.02             0.06             0.02             0.05  \n",
              "3             0.00             0.01             0.00             0.00  \n",
              "4             0.01             0.03             0.01             0.01  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-7870b209-0a4f-4403-b762-852b063963d5\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>CHROM</th>\n",
              "      <th>POS</th>\n",
              "      <th>REF</th>\n",
              "      <th>ALT</th>\n",
              "      <th>allele_count</th>\n",
              "      <th>allele_freq</th>\n",
              "      <th>eas_allele_freq</th>\n",
              "      <th>eur_allele_freq</th>\n",
              "      <th>afr_allele_freq</th>\n",
              "      <th>amr_allele_freq</th>\n",
              "      <th>sas_allele_freq</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>X</td>\n",
              "      <td>12583</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>97</td>\n",
              "      <td>0.02</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.07</td>\n",
              "      <td>0.01</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>X</td>\n",
              "      <td>13587</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>198</td>\n",
              "      <td>0.04</td>\n",
              "      <td>0.03</td>\n",
              "      <td>0.02</td>\n",
              "      <td>0.06</td>\n",
              "      <td>0.02</td>\n",
              "      <td>0.05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>X</td>\n",
              "      <td>13590</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>198</td>\n",
              "      <td>0.04</td>\n",
              "      <td>0.03</td>\n",
              "      <td>0.02</td>\n",
              "      <td>0.06</td>\n",
              "      <td>0.02</td>\n",
              "      <td>0.05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>X</td>\n",
              "      <td>13615</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>20</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.01</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.00</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>X</td>\n",
              "      <td>13878</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>66</td>\n",
              "      <td>0.01</td>\n",
              "      <td>0.00</td>\n",
              "      <td>0.01</td>\n",
              "      <td>0.03</td>\n",
              "      <td>0.01</td>\n",
              "      <td>0.01</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-7870b209-0a4f-4403-b762-852b063963d5')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-7870b209-0a4f-4403-b762-852b063963d5 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-7870b209-0a4f-4403-b762-852b063963d5');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#format VCF file (breakdown alleles)\n",
        "\n",
        "id_variant_list = []\n",
        "for id in tqdm(vcf.columns[9:]):\n",
        "  id_a = [id, 'A'] #first chromosome\n",
        "  id_b = [id, 'B'] #second chromosome\n",
        "  id_column = vcf[id].tolist() #get list of variants for each id\n",
        "  for variant in id_column:\n",
        "    variant_a = variant.split('|')[0]\n",
        "    variant_b = variant.split('|')[1]\n",
        "    id_a.append(variant_a)\n",
        "    id_b.append(variant_b)\n",
        "  \n",
        "  id_variant_list.append(id_a)\n",
        "  id_variant_list.append(id_b)\n",
        "\n",
        "#adding gene ids into dataframe.\n",
        "column_names = ['ID', 'pair']\n",
        "column_names.extend( (vcf['CHROM'] + '_' +  vcf['POS'].astype(str) + '_' + vcf['REF'] + '_' + vcf['ALT']).tolist() )\n",
        "\n",
        "id_variant_df = pd.DataFrame(id_variant_list)\n",
        "id_variant_df.columns = column_names\n",
        "id_variant_df.head()\n",
        "\n",
        "#save table to GDrive\n",
        "id_variant_df.to_csv('/content/drive/MyDrive/artificial_genome_project/chrX_vcf_formatted.csv', index=False)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3aWKqQth0BrI",
        "outputId": "def80abb-19a5-446d-ba2f-aaacd09d898a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "100%|██████████| 2548/2548 [03:30<00:00, 12.11it/s]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "---"
      ],
      "metadata": {
        "id": "VDvajis_0b17"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Prepare populations tags for genomes\n",
        "sample_df = pd.read_csv('/content/drive/MyDrive/arificial_genome_project/igsr-1000 genomes on grch38.tsv.tsv', delimiter = '\\t')[['Sample name', 'Population code', 'Population name', 'Superpopulation code', 'Superpopulation name']]\n",
        "sample_df.head()"
      ],
      "metadata": {
        "id": "q5UQnvHZ0cZc"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}